apis/discoveryengine/v1alpha1/datastore_reference.go

```diff
@@ -110,72 +124,76 @@ func NewDiscoveryEngineDataStoreRef(ctx context.Context, reader client.Reader, o
 		return nil, fmt.Errorf("cannot resolve resource ID")
 	}
 
-	// Use approved External
-	externalRef := valueOf(obj.Status.ExternalRef)
-	if externalRef == "" {
-		id.External = asDiscoveryEngineDataStoreExternal(id.parent, resourceID)
-		return id, nil
+	id := &DiscoveryEngineDataStoreID{
+		CollectionLink: &CollectionLink{
+			ProjectAndLocation: &ProjectAndLocation{
+				ProjectID: projectID,
+				Location:  location,
+			},
+			Collection: collectionID,
+		},
+		DataStore: resourceID,
 	}
 
-	// Validate desired with actual
-	actualParent, actualResourceID, err := parseDiscoveryEngineDataStoreExternal(externalRef)
-	if err != nil {
-		return nil, err
-	}
-	if actualParent.ProjectID != projectID {
-		return nil, fmt.Errorf("spec.projectRef changed, expect %s, got %s", actualParent.ProjectID, projectID)
-	}
-	if actualParent.Location != location {
-		return nil, fmt.Errorf("spec.location changed, expect %s, got %s", actualParent.Location, location)
-	}
-	if actualResourceID != resourceID {
-		return nil, fmt.Errorf("cannot reset `metadata.name` or `spec.resourceID` to %s, since it has already assigned to %s",
-			resourceID, actualResourceID)
-	}
-	id.External = externalRef
-	id.parent = &DiscoveryEngineDataStoreParent{ProjectID: projectID, Location: location}
-	return id, nil
-}
-
-func (r *DiscoveryEngineDataStoreRef) Parent() (*DiscoveryEngineDataStoreParent, error) {
-	if r.parent != nil {
-		return r.parent, nil
-	}
-	if r.External != "" {
-		parent, _, err := parseDiscoveryEngineDataStoreExternal(r.External)
+	// Validate the status.externalRef, if set
+	externalRef := valueOf(obj.Status.ExternalRef)
+	if externalRef != "" {
+		// Validate desired with actual
+		statusID, err := ParseDiscoveryEngineDataStoreExternal(externalRef)
 		if err != nil {
 			return nil, err
 		}
-		return parent, nil
+		if statusID.String() != id.String() {
+			return nil, fmt.Errorf("cannot change object key after creation; status=%q, new=%q",
+				statusID.String(), id.String())
+		}
 	}
-	return nil, fmt.Errorf("DiscoveryEngineDataStoreRef not initialized from `NewDiscoveryEngineDataStoreRef` or `NormalizedExternal`")
+	return id, nil
 }
 
-type DiscoveryEngineDataStoreParent struct {
+// +k8s:deepcopy-gen=false
```

Comment:
I'm a little bit get lost here. Is this because the parent is not used as  runtime.Object?

Reply:
Right, we don't need the deepcopy.

I'll remove, it's pretty harmless having an extra DeepCopy method.